### PR TITLE
Fix, document, and test parser and pretty-printer edge cases related to braced macro calls

### PR DESCRIPTION
**Pull request by dtolnay**
_Review note: this is a deceptively small PR because it comes with 145 lines of docs and 196 lines of tests, and only 25 lines of compiler code changed. However, I recommend reviewing it 1 commit at a time because much of the effect of the code changes is non-local i.e. affecting code that is not visible in the final state of the PR. I have paid attention that reviewing the PR one commit at a time is as easy as I can make it. All of the code you need to know about is touched in those commits, even if some of those changes disappear by the end of the stack._

This is a follow-up to https://github.com/rust-lang/rust/pull/119105. One case that is not relevant to `-Zunpretty=expanded`, but which came up as I'm porting #119105 and #118726 into `syn`'s printer and `prettyplease`'s printer where it **is** relevant, and is also relevant to rustc's `stringify!`, is statement boundaries in the vicinity of braced macro calls.

Rustc's AST pretty-printer produces invalid syntax for statements that begin with a braced macro call:

```rust
macro_rules! stringify_item {
    ($i:item) => {
        stringify!($i)
    };
}

macro_rules! repro {
    ($e:expr) => {
        stringify_item!(fn main() { $e + 1; })
    };
}

fn main() {
    println!("{}", repro!(m! {}));
}
```

**Before this PR:** output is not valid Rust syntax.

```console
fn main() { m! {} + 1; }
```

```console
error: leading `+` is not supported
 --> <anon>:1:19
  |
1 | fn main() { m! {} + 1; }
  |                   ^ unexpected `+`
  |
help: try removing the `+`
  |
1 - fn main() { m! {} + 1; }
1 + fn main() { m! {}  1; }
  |
```

**After this PR:** valid syntax.

```console
fn main() { (m! {}) + 1; }
```

### CODE REVIEW

File: compiler/rustc_ast/src/util/classify.rs
Comment (WaffleLapkin) [inline compiler/rustc_ast/src/util/classify.rs:59]: why is this needed, if all the call sites are not calling this if `e ≈ MacCall`?

File: tests/ui/macros/stringify.rs
Comment (WaffleLapkin) [inline tests/ui/macros/stringify.rs:229]: Can we correctly check to not add them?

File: compiler/rustc_parse/src/parser/expr.rs
Comment (WaffleLapkin) [inline compiler/rustc_parse/src/parser/expr.rs:501]: I'm not found of those `match`es. Can we instead have two `classify::` functions, one for match arms and one for other cases?

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]
I think I agree with the idea, but I'm not sure about the implementation. Is there a way to make this nicer?

**Comment by WaffleLapkin** [comment]
r? compiler

**Comment by fmease** [comment]
r? compiler

**Comment by dtolnay** [comment]
This is waiting on me to incorporate Waffle's insightful feedback, and to rebase.

File: tests/ui/macros/stringify.rs
Comment (dtolnay) [reply tests/ui/macros/stringify.rs]: Done. I found an appealing way to solve this, requiring #124191.

File: compiler/rustc_parse/src/parser/expr.rs
Comment (dtolnay) [reply compiler/rustc_parse/src/parser/expr.rs]: Good call. I've separated out `classify::expr_requires_semi_to_be_stmt` and `classify::expr_requires_comma_to_be_match_arm`, and there are no more matches like this.

File: compiler/rustc_ast/src/util/classify.rs
Comment (dtolnay) [reply compiler/rustc_ast/src/util/classify.rs]: In the commit that adds this `MacCall` case inside `expr_requires_semi_to_be_stmt`, I change all the call sites to preserve their pre-existing behavior. So this case starts out as not reachable. But after that, one at a time for each call site, I update each call site that *should* be reaching this, together with adding documentation and tests for that call site.

**Comment by michaelwoerister** [comment]
Thanks for the PR, @dtolnay, looks great! However, after looking through all the commits, I think I don't understand enough about the parsing subtleties involved here to review this.

r? parser

**Comment by dtolnay** [comment]
@compiler-errors recently reviewed the precursor PR (#124191) without even being assigned it, so they might have the easiest time with this one.

**Comment by dtolnay** [comment]
- Rebased to fix conflict with https://github.com/rust-lang/rust/pull/104087

File: compiler/rustc_parse/src/parser/expr.rs
Comment (dtolnay) [reply compiler/rustc_parse/src/parser/expr.rs:540]: Right, this is preserving old behavior. The exact same behavior that used to be incorrectly called `expr_requires_semi_to_be_stmt` is now called `expr_requires_comma_to_be_match_arm`.

Regarding naming, one possibility is that instead of distinguishing a "commas in match arms" and "semicolons in statements" case (and the implicit 3rd case, expressions that are neither statement nor match arm, such as a function arg), we could rename things along the following lines:

- "parse the longest possible expression"
- "parse an expression using earlier boundary rule" (for a match arm, range, or whatever else looks at expr_is_complete)
- "parse an expression until the earliest possible boundary" (for a statement)

Another possibility is to omit this specific commit (_"Add classify::expr_requires_comma_to_be_match_arm"_) and keep using `match e { MacCall(_) => …, _ => expr_requires_semi_to_be_stmt(e) }` in the cases where `expr_requires_semi_to_be_stmt` is not the desired behavior.

Comment (dtolnay) [reply compiler/rustc_parse/src/parser/expr.rs:540]: One more option &mdash; instead of having what's currently in the PR:

```rust
// compiler/rustc_ast/src/util/classify.rs

pub fn expr_requires_semi_to_be_stmt(e: &ast::Expr) -> bool {
    match &e.kind {
        If(..)
        | Match(..)
        | Block(..)
        | While(..)
        | Loop(..)
        | ForLoop { .. }
        | TryBlock(..)
        | ConstBlock(..) => false,

        MacCall(mac_call) => mac_call.args.delim != Delimiter::Brace,

        _ => true,
    }
}

pub fn expr_requires_comma_to_be_match_arm(e: &ast::Expr) -> bool {
    match &e.kind {
        MacCall(_) => true,
        _ => expr_requires_semi_to_be_stmt(e),
    }
}
```

we could have:

```rust
pub fn expr_is_complete(e: &ast::Expr) -> bool {
    // this is negative of the old "expr_requires_semi_to_be_stmt"
    match &e.kind {
        If(..)
        | Match(..)
        | Block(..)
        | While(..)
        | Loop(..)
        | ForLoop { .. }
        | TryBlock(..)
        | ConstBlock(..) => true,

        MacCall(_) | _ => false,
    }
}

pub fn expr_requires_semi_to_be_stmt(e: &ast::Expr) -> bool {
    match &e.kind {
        MacCall(mac_call) => mac_call.args.delim != Delimiter::Brace,
        _ => !expr_is_complete(e),
    }
}
```

Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/expr.rs:540]: I prefer either the final option or the renaming you suggested first above. I'll leave it up to you. Otherwise this PR looks good to me.

Comment (dtolnay) [reply compiler/rustc_parse/src/parser/expr.rs]: I removed https://github.com/rust-lang/rust/commit/5d5954c74df2b6fdfc89a29b502bb898a51fcd1d and added https://github.com/rust-lang/rust/commit/f895205d41ced8ddc7b9f72fdf9e474292aa2cb8.

**Review by compiler-errors — APPROVED** [review]
🆒

**Pull request merged**
